### PR TITLE
feat: Set mg_.. fields ( insert/update on/by) if passed in post

### PR DESCRIPTION
**Pull request by connoratrug**
- If the passed data includes mg_... fields, use these values instaed of the values computed by the system

### What are the main changes you did

- on persist , check if mg_... values are passed , if so use passed values instead of computing the values 
- only manager /admin roles can pass mg_[insert|update][on|by] fields

### How to test

- post csv with mg_... header and values -> see values are set 
- post csv without mg_... header and values -> see computed by the system 

Note: the current frontend code includes the mg_.. fields , ie these are used instead of the computed system fields

### Checklist

- [ ] checked that code complies with the [dev guidelines](https://github.com/molgenis/molgenis-emx2/blob/master/docs/molgenis/dev_guidelines.md)
- [ ] frontend code follows good semantic HTML practices and meets accessibility guidelines [Accessibility guide](https://github.com/molgenis/molgenis-emx2/blob/master/docs/molgenis/dev_accessibility.md)
- [ ] updated docs in case of new feature
- [ ] added/updated tests
- [ ] added/updated testplan to include a test for this fix, including ref to bug using # notation

By creating this pull request I have agreed with the [contributor terms](https://github.com/molgenis/molgenis-emx2/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java [hunk 1]
```diff
@@ -368,8 +369,29 @@ private List<Record> insertBatch(
...
+              && insertedColumnNames.contains(insertMetadataColumn)
+              && allRowsProvide(rows, insertMetadataColumn)) {
+            step2.set(
+                field(name(insertMetadataColumn)), (Object) getExcludedField(insertMetadataColumn));
+          }
+        }
+        step2.set(
+            field(name(MG_UPDATEDBY)),
+            insertedColumnNames.contains(MG_UPDATEDBY)
+                ? (Object) getExcludedField(MG_UPDATEDBY)
```
Comment (jhhaanstra) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java:388]: These casts are redundant

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java [hunk 2]
```diff
@@ -434,6 +467,30 @@ private int updateBatch(SqlTable table, List<Row> rows, List<Column> updateColum
...
+  /**
+   * Keeps a value that was supplied in the row (e.g. mg_insertedBy on import), otherwise applies
+   * the default. Note that the key is only replaced, never added, so the value order keeps matching
+   * the insert fields.
+   */
+  private static void putMgValue(
+      Map<String, Object> values,
+      String key,
+      boolean mayOverrideMgValues,
+      Supplier<Object> defaultValue) {
```
Comment (jhhaanstra) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java:488]: I find that the supplier doesn't really add anything and only increases complexity. The lookups that are provided, don't interact with the database but simply look at attributes, so might as well provide the defaultValue directly.

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java [hunk 3]
```diff
@@ -368,8 +369,29 @@ private List<Record> insertBatch(
         step2.set(column.getJooqField(), (Object) getExcludedField(column));
       }
       if (!inherit) {
-        step2.set(field(name(MG_UPDATEDBY)), getActiveUser(table));
-        step2.set(field(name(MG_UPDATEDON)), now);
+        // the inserted values already hold either the row supplied value or the default,
```
Comment (jhhaanstra) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java:372]: This method is getting pretty large, can we extract this into a separate method with a good name?

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java [hunk 4]
```diff
@@ -434,6 +467,30 @@ private int updateBatch(SqlTable table, List<Row> rows, List<Column> updateColum
...
+    return rows.stream().allMatch(row -> row.notNull(columnName));
+  }
+
+  /** only admins and managers may decide what ends up in the mg_ metadata columns */
+  private static boolean mayOverrideMgValues(SqlTable table) {
+    return PermissionEvaluator.canManage(table.getSchema());
+  }
+
+  /**
+   * Keeps a value that was supplied in the row (e.g. mg_insertedBy on import), otherwise applies
```
Comment (jhhaanstra) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java:480]: This javadoc is not needed, the method is pretty self-explanatory and we need to maintain javadoc as well.

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java [hunk 5]
```diff
@@ -368,8 +369,29 @@ private List<Record> insertBatch(
         step2.set(column.getJooqField(), (Object) getExcludedField(column));
       }
       if (!inherit) {
-        step2.set(field(name(MG_UPDATEDBY)), getActiveUser(table));
-        step2.set(field(name(MG_UPDATEDON)), now);
+        // the inserted values already hold either the row supplied value or the default,
+        // so on conflict we reuse them; if the column is not part of the insert we apply the
+        // default
+        List<String> insertedColumnNames = columns.stream().map(Column::getName).toList();
+        // insert metadata of the existing row is only overwritten when every row supplies it
```
Comment (jhhaanstra) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java:376]:
```
// insert metadata of the existing row is only overwritten when every row supplies it
```
and when we have sufficient rights, and when it's part of `INSERT_METADATA_COLUMNS`.

We dont need this comment

File: backend/molgenis-emx2-sql/src/test/java/org/molgenis/emx2/sql/TestMgColumns.java [hunk 6]
```diff
@@ -6,130 +6,246 @@
...
-    assertEquals(inserted.getDateTime(MG_INSERTEDON), saved.getDateTime(MG_INSERTEDON));
-    assertEquals(inserted.getString(MG_INSERTEDBY), saved.getString(MG_INSERTEDBY));
+    assertInsertedMetadata(
+        saved, inserted.getString(MG_INSERTEDBY), inserted.getDateTime(MG_INSERTEDON));
     assertTrue(saved.getDateTime(MG_INSERTEDON).isBefore(saved.getDateTime(MG_UPDATEDON)));
   }
+
+  @Test
+  void testProvidedMgValuesAreUsedOnInsertUpdateAndSave() {
+    Table t = schema.create(table("ProvidedMg", column("id").setPkey(), column("value")));
```
Comment (jhhaanstra) [inline backend/molgenis-emx2-sql/src/test/java/org/molgenis/emx2/sql/TestMgColumns.java:152]: rename t -> table

A while ago in the dev meeting I thought we agreed upon not using abbreviations as var names anymore (22nd july)

Can you apply this to the rest of the class as well?

File: backend/molgenis-emx2-sql/src/test/java/org/molgenis/emx2/sql/TestMgColumns.java
Comment (jhhaanstra) [inline backend/molgenis-emx2-sql/src/test/java/org/molgenis/emx2/sql/TestMgColumns.java:1]: I miss a test case where you check the following edge case:

given multiple rows, if one doesn't have an MG column, while the others have, you don't override anything.

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java [hunk 7]
```diff
@@ -368,8 +369,29 @@ private List<Record> insertBatch(
...
+          if (mayOverrideMgValues
+              && insertedColumnNames.contains(insertMetadataColumn)
+              && allRowsProvide(rows, insertMetadataColumn)) {
+            step2.set(
+                field(name(insertMetadataColumn)), (Object) getExcludedField(insertMetadataColumn));
+          }
+        }
+        step2.set(
+            field(name(MG_UPDATEDBY)),
+            insertedColumnNames.contains(MG_UPDATEDBY)
```
Comment (jhhaanstra) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java:387]: At this point in the pipeline, we always have the MG fields present in the rows that we want to insert. That is because of `SqlTable#getInsertColumns`, which is always used for input. I'm not sure on whether we want to remove this or not, but with the current setup it is redundant.

If we remove it, we get tight coupling with getInsertColumns, but leaving it also increases the complexity of this method

**Review by jhhaanstra — CHANGES_REQUESTED** [review]

**Review by mswertz — COMMENTED** [review]
looks good, happy for both of you to finish it together.

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java (hunk 7 above)
Comment (connoratrug) [reply backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java]: no sure what your suggesting here

**Review by dtroelofsprins — APPROVED** [review]
Functional test:

Tested several scenarios included in one file (empty mg_field columns, updated InsertedOn/By and updatedOn/By. Also the initial purpose, if you have a large model update and need to do all out, all in, now makes it possible to preserve the original insertedOn and mgUpdatedOn dates.

It works also fine for the mg_roles column

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java [hunk 8]
```diff
@@ -318,62 +317,116 @@ private TableListener getTableListener() {
...
+    onConflict.set(
+        field(name(MG_UPDATEDON)),
+        insertedColumnNames.contains(MG_UPDATEDON)
+            ? getExcludedField(MG_UPDATEDON)
+            : mgDefaults.now());
+  }
+
+  private record MgDefaults(String user, LocalDateTime now, boolean mayOverride) {
+    static MgDefaults of(SqlTable table) {
+      return new MgDefaults(getActiveUser(table), LocalDateTime.now(), mayOverrideMgValues(table));
```
Comment (jhhaanstra) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java:421]: Do we want to do something with this sonar warning?

Comment (connoratrug) [reply backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTable.java:421]: warning ?

**Review by jhhaanstra — APPROVED** [review]
